### PR TITLE
[YUNIKORN-552] Add some unit tests to cover placeholder cleanup

### PR DESCRIPTION
**Pull request by kobe860219**
### What is this PR for?
Before this PR, the [code](https://github.com/apache/incubator-yunikorn-k8shim/blob/9df2f67e3d955a5d2de36cc0540cc86268a9d683/pkg/cache/placeholder_manager.go#L146) has not be tested yet. Add an unit test for up it test coverage.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-552

### How should this be tested?
1. go test ./pkg/cache -cover -race -tags deadlock -coverprofile=coverage.txt -covermode=atomic
2. go tool cover -html=coverage.txt

### Screenshots (if appropriate)
image

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by kobe860219** [comment]
@yangwwei  @HuangTing-Yao @0yukali0 Please help me review this pr. Thanks !

File: pkg/cache/placeholder_manager_test.go [hunk 1]
```diff
@@ -261,3 +261,24 @@ func TestPlaceholderManagerStartStop(t *testing.T) {
...
+			Name: "pod-01",
+			UID:  "UID-01",
+		},
+	}
+	mgr := NewPlaceholderManager(mockedAPIProvider.GetAPIs())
+	mgr.Start()
+	assert.Equal(t, mgr.isRunning(), true, "manager should be running after start")
+	mgr.orphanPods["task01"] = pod1
+	assert.Equal(t, len(mgr.orphanPods), 1)
+	<-time.After(5 * time.Second)
```
Comment (yangwwei) [inline pkg/cache/placeholder_manager_test.go:282]: we can't wait for 5s in UT for this single test
what we can do here is to add a function that can change the default 5s interval for the placeholder cleanup, e.g `setCleanupInterval(interval time.Duration)`. During this test, set the interval to 100millsecond, etc. And remember to set it back to 5s after the test.

File: pkg/cache/placeholder_manager_test.go [hunk 2]
```diff
@@ -261,3 +261,24 @@ func TestPlaceholderManagerStartStop(t *testing.T) {
...
+		},
+		ObjectMeta: apis.ObjectMeta{
+			Name: "pod-01",
+			UID:  "UID-01",
+		},
+	}
+	mgr := NewPlaceholderManager(mockedAPIProvider.GetAPIs())
+	mgr.Start()
+	assert.Equal(t, mgr.isRunning(), true, "manager should be running after start")
+	mgr.orphanPods["task01"] = pod1
```
Comment (yangwwei) [inline pkg/cache/placeholder_manager_test.go:280]: can we add more than 1 pod here?

**Review by yangwwei — CHANGES_REQUESTED** [review]

**Comment by kobe860219** [comment]
image

File: pkg/cache/placeholder_manager_test.go (hunk 2 above)
Comment (kobe860219) [reply pkg/cache/placeholder_manager_test.go:291]: image

File: pkg/cache/placeholder_manager_test.go (hunk 1 above)
Comment (kobe860219) [reply pkg/cache/placeholder_manager_test.go]: image

File: pkg/cache/placeholder_manager_test.go (hunk 2 above)
Comment (kobe860219) [reply pkg/cache/placeholder_manager_test.go:291]: Added

File: pkg/cache/placeholder_manager_test.go [hunk 3]
```diff
@@ -261,3 +261,24 @@ func TestPlaceholderManagerStartStop(t *testing.T) {
...
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: apis.ObjectMeta{
+			Name: "pod-01",
+			UID:  "UID-01",
+		},
+	}
+	mgr := NewPlaceholderManager(mockedAPIProvider.GetAPIs())
+	mgr.Start()
```
[the remark this answers is not in the document]
Comment (kobe860219) [reply pkg/cache/placeholder_manager_test.go:289]: Added

**Comment by 0yukali0** [comment]
Hi @kobe860219 
1.orphanPods is accessed by  cleanOrphanPlaceholders and len(mgr.orphanPods) same time.
You need a lock to protect this variable.
> func (mgr *PlaceholderManager) GetorphanPodsLength {
  mgr.Lock()
  defer mgr.Unlock()
  return len(mgr.orphanPods)
}

2.cleanupTime is also accessed  to write and read same time.
make sure this variable is accessed by only one process at one time.
> func (mgr *PlaceholderManager) GetCleanupTime {
  mgr.Lock()
  defer mgr.Unlock()
  return mgr.cleanupTime
}

func (mgr *PlaceholderManager) SetCleanupTime {
  mgr.Lock()
  defer mgr.Unlock()
  mgr.cleanupTime = 5 * time.Second
}

**Comment by kobe860219** [comment]
@yangwwei  Hi, Could you help me review it. Thanks!

**Review by kingamarton — CHANGES_REQUESTED** [review]

File: pkg/cache/placeholder_manager_test.go [hunk 4]
```diff
@@ -261,3 +261,41 @@ func TestPlaceholderManagerStartStop(t *testing.T) {
...
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: apis.ObjectMeta{
+			Name: "pod-02",
+			UID:  "UID-02",
+		},
+	}
+	mgr := NewPlaceholderManager(mockedAPIProvider.GetAPIs())
+	mgr.setCleanupTime(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (kobe860219) [reply pkg/cache/placeholder_manager_test.go:288]: Thanks ! This tip is very helpful for me !

File: pkg/cache/placeholder_manager.go [hunk 5]
```diff
@@ -165,3 +167,16 @@ func (mgr *PlaceholderManager) isRunning() bool {
 func (mgr *PlaceholderManager) setRunning(flag bool) {
 	mgr.running.Store(flag)
 }
+
+func (mgr *PlaceholderManager) getOrphanPodsLength() int {
+	mgr.Lock()
+	defer mgr.Unlock()
+	orphanPodsLength := len(mgr.orphanPods)
```
[the remark this answers is not in the document]
Comment (kobe860219) [reply pkg/cache/placeholder_manager.go]: I will remove unessential variable.

File: pkg/cache/placeholder_manager.go [hunk 6]
```diff
@@ -165,3 +167,16 @@ func (mgr *PlaceholderManager) isRunning() bool {
 func (mgr *PlaceholderManager) setRunning(flag bool) {
 	mgr.running.Store(flag)
 }
+
+func (mgr *PlaceholderManager) getOrphanPodsLength() int {
+	mgr.Lock()
+	defer mgr.Unlock()
```
[the remark this answers is not in the document]
Comment (kobe860219) [reply pkg/cache/placeholder_manager.go]: Thanks for your review. I replaced `Mute` with `RWMute` for `RLock()` in `PlaceholderManager` struct.

**Comment by kobe860219** [comment]
@yangwwei  Hi, thanks your help for rerun the check. The e2e test now is ok !

**Review by wilfred-s — APPROVED** [review]
LGTM thank you for the additional tests

**Pull request closed without merging**